Refactoring and other improvements

shankari opened the pull request:
Refactor code minimally to make sense and to avoid duplicated code.
Avoiding significant functional restructuring due to lack of unit tests.
Focus is on (minimal) readability and performance.

shankari committed:
Simple changes towards standardization before the big refactoring

- Replace the participant filter and the direct call to `edb` with
    the standard call to the abstract timeseries

- Change the imports to the standard format with standard imports first, and
  our imports later

shankari committed:
Merge branch 'modeling_and_functions' of https://github.com/corinne-hcr/e-mission-server into modeling_and_functions

shankari committed:
Improve the implementation of `get_all_uuids()`

The previous implementation was in a situation where we didn't know for sure
whether or not there would be UUID entries for each user. But that has been
true by default for years now.

If we ever got into a situation in which we didn't have the uuid_db, we would
just create the missing entries.

Switching to the UUID DB avoids weird UUIDs like `None` and `null`, and also
is significantly faster.

```
%timeit edb.get_uuid_db().distinct("uuid")
1.29 ms ± 12 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

```
%timeit edb.get_timeseries_db().distinct("user_id")
3.68 s ± 105 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
```

shankari committed:
Add logging + minor improvements to make code "pythonic"

- Enable logging
- Add logging for each stage of the pipeline
- change all `for i in range(len(x))` to `for y in x` or `for i, y in
  enumerate(x)` as needed
- Change the logging in `map_labels_mode` to log all the rows instead of one
  row at a time, to be consistent with the log message

shankari commented:
After logging, the time taken for each step is:

```
2021-07-19 21:43:11,376:INFO:At stage: Reading data
2021-07-19 21:43:11,436:INFO:At stage: Filtering data
2021-07-19 21:43:12,880:INFO:At stage: Splitting data
2021-07-19 21:43:12,930:INFO:At stage: starting tuning for stage 0
2021-07-19 21:44:10,586:INFO:At stage: starting tuning for stage 1
2021-07-19 21:44:58,866:INFO:At stage: starting tuning for stage 2
2021-07-19 21:45:41,131:INFO:At stage: starting tuning for stage 3
2021-07-19 21:46:32,471:INFO:At stage: starting tuning for stage 4
2021-07-19 21:47:35,710:INFO:At stage: starting testing for stage 0
2021-07-19 21:47:40,489:INFO:At stage: starting testing for stage 1
2021-07-19 21:47:45,269:INFO:At stage: starting testing for stage 2
2021-07-19 21:47:50,158:INFO:At stage: starting testing for stage 3
2021-07-19 21:47:55,039:INFO:At stage: starting testing for stage 4
2021-07-19 21:47:59,743:INFO:At stage: parameter selection outputs complete
```

The tuning stages take around a minute each and the testing stages take around 5 seconds each.

shankari commented:
Skimming through the logs, we appear to calculate the distance matrix multiple times

```
kshankar-35069s:gis_branch_tests kshankar$ grep "Considering trip 60053328db3d82eed967bdd9" /tmp/eval_pipeline_user1.log
2021-07-19 21:43:11,441:DEBUG:Considering trip 60053328db3d82eed967bdd9:
2021-07-19 21:46:32,471:DEBUG:Considering trip 60053328db3d82eed967bdd9:
2021-07-19 21:47:35,710:DEBUG:Considering trip 60053328db3d82eed967bdd9:
2021-07-19 21:47:40,489:DEBUG:Considering trip 60053328db3d82eed967bdd9:
2021-07-19 21:47:45,270:DEBUG:Considering trip 60053328db3d82eed967bdd9:
2021-07-19 21:47:50,159:DEBUG:Considering trip 60053328db3d82eed967bdd9:
```

and 

```
2021-07-19 21:43:11,447:DEBUG:endpoints are = (X, Y) and (A, B)
2021-07-19 21:46:32,477:DEBUG:endpoints are = (X, Y) and (A, B)
2021-07-19 21:47:35,716:DEBUG:endpoints are = (X, Y) and (A, B)
2021-07-19 21:47:40,494:DEBUG:endpoints are = (X, Y) and (A, B)
2021-07-19 21:47:45,275:DEBUG:endpoints are = (X, Y) and (A, B)
2021-07-19 21:47:50,163:DEBUG:endpoints are = (X, Y) and (A, B)
```

But refactoring to remove it seems like too much at this point

shankari commented:
Something that we should be able to change are the multiple calls to 

```
2021-07-19 21:43:12,930:INFO:At stage: starting tuning for stage 0
2021-07-19 21:43:13,436:DEBUG:The following rows will be changed Int64Index([], dtype='i
2021-07-19 21:43:13,437:DEBUG:Finished changing all rows
...
2021-07-19 21:44:10,582:DEBUG:The following rows will be changed Int64Index([], dtype='int64')
2021-07-19 21:44:10,583:DEBUG:Finished changing all rows
2021-07-19 21:44:10,586:INFO:At stage: starting tuning for stage 12021-07-19 21:44:10,586:INFO:At stage: starting tuning for stage 1
```

It takes close to a minute, and given the code which maps all the user inputs in one function, there is really no reason to invoke it **8050 times** for one tuning run. 

```
$ grep Finished /tmp/foo | wc -l
    8050

$ head -n 2 /tmp/foo
2021-07-19 21:43:13,436:DEBUG:The following rows will be changed Int64Index([], dtype='int64')
2021-07-19 21:43:13,437:DEBUG:Finished changing all rows
...
$ tail -n 2 /tmp/foo
2021-07-19 21:44:10,582:DEBUG:The following rows will be changed Int64Index([], dtype='int64')
2021-07-19 21:44:10,583:DEBUG:Finished changing all rows
```

```
 def map_labels_mode(user_input_df):
     # convert mode
+    same_mode_df = user_input_df[user_input_df.replaced_mode == "same_mode"]
+    logging.debug("The following rows will be changed %s" %
+        same_mode_df.index)
     for a in range(len(user_input_df)):
         if user_input_df.iloc[a]["replaced_mode"] == "same_mode":
             # to see which row will be converted
-            logging.debug("The following rows will be changed: %s", user_input_df.iloc[a])
+            # logging.debug("The following rows will be changed: %s", user_input_df.iloc[a])
             user_input_df.iloc[a]["replaced_mode"] = user_input_df.iloc[a]['mode_confirm']
+    logging.debug("Finished changing all rows")
     return user_input_df
```

shankari commented:
The multiple calls appear to come from 

```
$ grep "Calculating score" /tmp/foo | wc -l
    8050
```

```
def score(bin_trips, labels_pred):
    bin_trips_user_input_df = pd.DataFrame(data=[trip["data"]["user_input"] for trip in bin_trips])
    bin_trips_user_input_df = label_pro.map_labels(bin_trips_user_input_df)
    # turn all user_input into list without binning
    bin_trips_user_input_ls = bin_trips_user_input_df.values.tolist()
```

This is the problem with not using dataframes consistently - you pass in the data as a list of trips, convert it to a dataframe and map the labels in the dataframe **every time** you calculate the score. And then you turn it into a list anyway.

But why are we calling `score` 8050 times?

shankari commented:
Ah, it is because of 

```
        for dist_pct in np.arange(0.15, 0.6, 0.02):
            for low in range(250, 600):

                percentage_second, homo_second = second_round(bin_trips,filter_trips,first_labels,track,low,dist_pct,
                                                              sim,kmeans)

                curr_score = gs.get_score(homo_second, percentage_second)
                if curr_score not in tune_score:
                    tune_score[curr_score] = (low, dist_pct)
```

So calling `score` 8050 times is legit, because we call `second_round` 8050 times. We can definitely improve the implementation of this (list -> dataframe -> list), but it is not super egregious, so let's move on and refactor after we get everything working end-to-end.

shankari committed:
Improve logging in `get_scores` and `similarity`

shankari commented:
At this point, we can generate the tuning result for a user. Let's move on and make sure that we can run `build_save_model`.

shankari commented:
Error:

```
  File "emission/analysis/modelling/tour_model/build_save_model.py", line 153, in <module>
    main()
  File "emission/analysis/modelling/tour_model/build_save_model.py", line 116, in main
    unique_labels['p'] = unique_labels.count / sum_trips
```

shankari committed:
Minor fixes to be able to run the code

+ enable logging
+ replace `count` with `uniqcount` so it doesn't conflict with the count() function
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.count.html

Without this fix, we run into:
https://github.com/corinne-hcr/e-mission-server/pull/2#issuecomment-883086039

shankari committed:
Add logging to the final step of the pipeline as well

shankari committed:
Bump up the list of userids that the pipelines run on

Instead of running on only the first one

shankari committed:
:ta-da Integrate the label prediction into the inference pipeline

This required some changes to the intake pipeline code:
- add a new algorithm type
- add a new non-placeholder function that calls the actual load prediction code

This also required changes to the load_predict code to match the intake pipeline:
- accept a single trip instead of an array as input
    - change the trip access functions accordingly
- enable logging + add additional log statements to know what we are doing
- convert the `print` statements in the ad-hoc unit testing to assert
  statements so we could actually see if they failed
- improve exception handling by catching and printing the exception

Testing done:
- Ran the intake pipeline; it passed
- Connected a UI client from master, checked that there were no errors

shankari commented:
At the current point, everything seems to work end-to-end for the minipilot dataset.

The labeling step does not take very long; in fact, it takes substantially less time that trip segmentation.

```
2021-07-20T00:51:41.819742-07:00**********UUID 1044195f-af9e-43d4-9407-60594e5e9938: filter accuracy if needed**********
2021-07-20T00:51:41.831047-07:00**********UUID 1044195f-af9e-43d4-9407-60594e5e9938: segmenting into trips**********
2021-07-20T01:02:51.599526-07:00**********UUID 1044195f-af9e-43d4-9407-60594e5e9938: segmenting into sections**********
2021-07-20T01:03:36.753530-07:00**********UUID 1044195f-af9e-43d4-9407-60594e5e9938: smoothing sections**********
2021-07-20T01:04:05.819435-07:00**********UUID 1044195f-af9e-43d4-9407-60594e5e9938: cleaning and resampling timeline**********
/Users/kshankar/miniconda-4.8.3/envs/emission/lib/python3.7/site-packages/numpy/lib/nanfunctions.py:1111: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
/Users/kshankar/miniconda-4.8.3/envs/emission/lib/python3.7/site-packages/numpy/lib/nanfunctions.py:1111: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
/Users/kshankar/miniconda-4.8.3/envs/emission/lib/python3.7/site-packages/numpy/lib/nanfunctions.py:1111: RuntimeWarning: Mean of empty slice
  return np.nanmean(a, axis, out=out, keepdims=keepdims)
2021-07-20T01:07:27.461055-07:00**********UUID 1044195f-af9e-43d4-9407-60594e5e9938: inferring transportation mode**********
Error while inferring modes, timestamp is unchanged
Traceback (most recent call last):
  File "/Users/kshankar/e-mission/gis_branch_tests/emission/analysis/classification/inference/mode/pipeline.py", line 41, in predict_mode
    mip.runPredictionPipeline(user_id, time_query)
  File "/Users/kshankar/e-mission/gis_branch_tests/emission/analysis/classification/inference/mode/pipeline.py", line 139, in runPredictionPipeline
    self.loadModelStage()
  File "/Users/kshankar/e-mission/gis_branch_tests/emission/analysis/classification/inference/mode/pipeline.py", line 156, in loadModelStage
    self.model = seedp.ModeInferencePipelineMovesFormat.loadModel()
  File "/Users/kshankar/e-mission/gis_branch_tests/emission/analysis/classification/inference/mode/seed/pipeline.py", line 93, in loadModel
    fd = open(SAVED_MODEL_FILENAME, "r")
FileNotFoundError: [Errno 2] No such file or directory: 'seed_model.json'
2021-07-20T01:07:27.666963-07:00**********UUID 1044195f-af9e-43d4-9407-60594e5e9938: inferring labels**********
2021-07-20T01:07:39.791206-07:00**********UUID 1044195f-af9e-43d4-9407-60594e5e9938: populating expectations**********
2021-07-20T01:07:41.131361-07:00**********UUID 1044195f-af9e-43d4-9407-60594e5e9938: creating confirmed objects **********
2021-07-20T01:08:02.150297-07:00**********UUID 1044195f-af9e-43d4-9407-60594e5e9938: storing views to cache**********
```

shankari commented:
And the UI is populated with at least some values.

![Screen Recording 2021-07-20 at 1 24 46 AM mov](https://user-images.githubusercontent.com/2423263/126288504-a2b2faf7-9a59-4531-bec5-55db75419808.gif)

shankari commented:
@GabrielKS, at least on a Galaxy Nexus 720x1280 hdpi emulator, the top button bar is all messed up (see above).

shankari commented:
Since the tuning is likely to take many hours to complete, I will deploy this unchanged on staging and launch the evaluation pipeline. Then we can continue end-to-end testing tomorrow morning.

shankari commented:
Tuning is complete. 35 of the users are invalid.

```
# grep "warn.*early return" /tmp/eval_pipeline_all.log | wc -l
35
```

4 are valid

```
# ls user_*.csv | wc -l
4
```

shankari commented:
Let's build the model for those 4 users now.

shankari commented:
Building the model, there are still 35 invalid users

```
# grep "This user doesn't have" /tmp/build_model_all.log | wc -l
35
```

but only 3 valid users

```
# ls -1 models_* | wc -l
3
```

shankari commented:
Looks like that user does have some entries, but they are all individual trips

```
2021-07-20 13:06:49,622:DEBUG:for user ..., returning timeseries <emission.storage.timeseries.builtin_timeseries.BuiltinTimeSeries object at 0x7fb42128d890>
2021-07-20 13:06:49,809:DEBUG:After removing trips that are points, there are 18 data points
...
2021-07-20 13:06:49,836:DEBUG:bins = [[0], [1], [2], [3], [4], [5], [6], [7], [8], [9], [10], [11], [12], [13], [14], [15], [16], [17]], elbow distance = -1
2021-07-20 13:06:49,836:DEBUG:the new number of trips is 0
2021-07-20 13:06:49,836:DEBUG:the cutoff point is 0
```

And it is the only one with that characteristic

```
# grep "the cutoff point" /tmp/build_model_all.log
2021-07-20 13:06:42,988:DEBUG:the cutoff point is 1
2021-07-20 13:06:47,826:DEBUG:the cutoff point is 10
2021-07-20 13:06:49,836:DEBUG:the cutoff point is 0
2021-07-20 13:06:51,517:DEBUG:the cutoff point is 7
```

shankari commented:
Looking at the output tuning parameters for this user, they are all "basic"

```
0,user32,X,1,1,1,,0.5,0,0,1.0
1,user32,X,1,1,1,,0.5,0,0,1.0
2,user32,X,1,1,1,,0.5,0,0,1.0
3,user32,X,1,1,1,,0.5,0,0,1.0
4,user32,X,1,1,1,,0.5,0,0,1.0
```

But looking at the cutoff points, there appear to be multiple users for whom all the cutoffs are 0. Why is only this one saved?

shankari commented:
Looking at the zero cutoffs during parameter selection, many of them were for users where some of the splits had a low cutoff. Focusing on the first user, they has zero cutoffs for all the tuning data, but non-zero data for the testing cutoffs.

```
2021-07-20 08:46:32,619:DEBUG:the cutoff point is 0
2021-07-20 08:46:32,620:INFO:At stage: starting tuning for stage 1
--
2021-07-20 08:46:32,636:DEBUG:the cutoff point is 0
2021-07-20 08:46:32,637:INFO:At stage: starting tuning for stage 2
--
2021-07-20 08:46:32,653:DEBUG:the cutoff point is 0
2021-07-20 08:46:32,654:INFO:At stage: starting tuning for stage 3
--
2021-07-20 08:46:32,672:DEBUG:the cutoff point is 0
2021-07-20 08:46:32,672:INFO:At stage: starting tuning for stage 4
--
2021-07-20 08:46:32,690:DEBUG:the cutoff point is 0
2021-07-20 08:46:32,691:INFO:At stage: starting testing for stage 0
```

But for user X, both the tuning and testing cutoffs were 0

```
2021-07-20 08:52:20,114:DEBUG:the cutoff point is 0
2021-07-20 08:52:20,115:INFO:At stage: starting tuning for stage 1
--
2021-07-20 08:52:20,127:DEBUG:the cutoff point is 0
2021-07-20 08:52:20,128:INFO:At stage: starting tuning for stage 2
--
2021-07-20 08:52:20,141:DEBUG:the cutoff point is 0
2021-07-20 08:52:20,142:INFO:At stage: starting tuning for stage 3
--
2021-07-20 08:52:20,152:DEBUG:the cutoff point is 0
2021-07-20 08:52:20,154:INFO:At stage: starting tuning for stage 4
--
2021-07-20 08:52:20,167:DEBUG:the cutoff point is 0
2021-07-20 08:52:20,168:INFO:At stage: starting testing for stage 0
--
2021-07-20 08:52:20,230:DEBUG:the cutoff point is 0
2021-07-20 08:52:20,232:INFO:At stage: starting testing for stage 1
--
2021-07-20 08:52:20,301:DEBUG:the cutoff point is 0
2021-07-20 08:52:20,302:INFO:At stage: starting testing for stage 2
--
2021-07-20 08:52:20,376:DEBUG:the cutoff point is 0
2021-07-20 08:52:20,378:INFO:At stage: starting testing for stage 3
--
2021-07-20 08:52:20,436:DEBUG:the cutoff point is 0
2021-07-20 08:52:20,438:INFO:At stage: starting testing for stage 4
--
2021-07-20 08:52:20,514:DEBUG:the cutoff point is 0
2021-07-20 08:52:20,516:INFO:At stage: parameter selection outputs complete
```

So the selected split also has cutoff 0 and we are not able to build a model. I am not sure why we should build the final model based only on the selected split, though. Once we have generated the tuning parameters through cross-validation, we should  be able to train on the entire model, since we will be predicting values on new data, which is by definition, not part of the training set.

shankari commented:
Anyway, plowing ahead to at least get some predictions for the three valid users for the demo today...
The regular intake pipeline should just fill in labels for any new trips. Let's check that the pipeline is not generating errors for users who don't have a saved model.

```
>>> edb.get_pipeline_state_db().find({"pipeline_stage": 14, "last_processed_ts": {"$ne": None}}).count()
35
```

Ok, so there are 35 successful runs of pipeline stage 14 (`LABEL_INFERENCE`).
The 4 missing entries don't overlap with the entries for which we have models.

shankari commented:
Checking the entries for which we have models to see if they ran successfully...

```
2021-07-20 09:15:19,942:INFO:139993213572928:**********UUID X: inferring labels**********
2021-07-20 09:15:19,992:DEBUG:139993213572928:orig_ts_db_matches = 0, analysis_ts_db_matches = 54
2021-07-20 09:15:20,050:ERROR:139993213572928:[Errno 2] No such file or directory: 'locations_X'

2021-07-20 10:15:20,376:INFO:140093910353728:**********UUID X: inferring labels**********
2021-07-20 10:15:20,398:DEBUG:140093910353728:orig_ts_db_matches = 0, analysis_ts_db_matches = 0
```

When we first ran the pipeline, the models were not yet built, so we weren't able to predict. But when we are unable to load models, we don't actually fail the step, we return empty. This means that even when the models were available, we thought that the predictions were up-to-date, and didn't predict anything.

Resetting the pipeline state for `LABEL_INFERENCE` (14) and `EXPECTATION_POPULATION` (15) for those three users now...

shankari commented:
Reset for all users to make life simpler, pipeline will run again at 15 mins past the hour

```
>>> edb.get_analysis_timeseries_db().delete_many({"metadata.key": "analysis/inferred_trip"})
<pymongo.results.DeleteResult object at 0x7f122d436230>
>>> edb.get_analysis_timeseries_db().delete_many({"metadata.key": "analysis/expected_trip"})
<pymongo.results.DeleteResult object at 0x7f122c1ce5a0>
>>> edb.get_pipeline_state_db().delete_many({"pipeline_stage": {"$in": [14,15]}})
<pymongo.results.DeleteResult object at 0x7f122b929b40>
```

shankari commented:
Manually ran the pipeline for user `X`. 

Out of 54 trips, 3 were labeled

```
2021-07-20 15:43:22,187:DEBUG:139888701101888:orig_ts_db_matches = 0, analysis_ts_db_matches = 54
```

3 had inferred labels

```
# grep "second round labeling" /var/tmp/intake_single.log
2021-07-20 15:43:22,279:DEBUG:139888701101888:At stage: second round labeling
2021-07-20 15:43:22,380:DEBUG:139888701101888:At stage: second round labeling
2021-07-20 15:43:22,429:DEBUG:139888701101888:At stage: second round labeling
```

shankari commented:
The inferred labels were from May and June, and on spot checking, at least one of them had already been labeled. So it is not clear whether we will be able to see any inferred labels.

So for the demo, we will reset the `LABEL_INFERENCE (14)` and `EXPECTATION_POPULATION (15)` and `CREATE_CONFIRMED_OBJECTS (13)` stages, and change the inference to use a placeholder that selects from the previous label combinations from that user.

After the demo/improvements to the labeling inference algorithm, we will reset all of these again and see what we get in a week.

shankari commented:
The pipeline resets have a small problem - if we reset everything at one time, then the inference algorithms won't be able to build any models since they don't have any confirmed trips. So we need to ensure that we have confirmed trips, then reset the label inference only and re-run.

So our pipeline reset now (when we have confirmed_trips without inferences) would be:
- reset label inference
- run pipeline (will generate inferred trips)
- reset confirmed trips
- run pipeline (will generate confirmed trips with inferred labels)

GabrielKS commented:
> @GabrielKS, at least on a Galaxy Nexus 720x1280 hdpi emulator, the top button bar is all messed up (see above).

That's unfortunate. I've been testing on an iPhone 8 emulator; I'll have to do some Android tests when I can.

shankari committed:
New demo placeholder + label specific reset

We have a cold start problem with demos - we want to be able to showcase the
labeling, but it won't work until we have sufficient labels.

This commit introduces a new placeholder that will help with demos. It pulls
the unique set of existing labels and randomly returns one of them 9/10th of
the time. The remaining 1/10th of the time, it returns no inference.

This lets us prototype our goal, where most of the trips are labeled, and the
user only occasionally has to input labels.

However, if the user does have sufficiently good labels, we do want to use
that.  So we use our built-in ensemble method to generate two labels, pick the
"real", non-placeholder if if exists, and fallback to the placeholder if it
doesn't.

Note that if we enhanced the placeholder to use the actual distribution of the
existing labels, this could arguably be a very simple randomized algorithm.

+ added a super simple script to reset only the label inference stage and the
only the confirmed trip stage so that we can experiment with this without
having to re-run the complete pipeline

+ added a super simple script to print out label stats - how many of each kind of trip, to debug pipeline operation

Testing done:
- Reset the inference pipeline
- Ran the intake pipeline

Placeholder was invoked

```
2021-07-20 10:54:54,758:DEBUG:4457897408:In placeholder_predictor_demo: ound 85 for user 1d292b85-c549-409a-a10d-746e957582a0, returning value {'mode_confirm': 'shared_ride', 'purpose_confirm': 'meal', 'replaced_mode': 'same_mode'}
2021-07-20 10:54:54,866:DEBUG:4457897408:In placeholder_predictor_demo: ound 85 for user 1d292b85-c549-409a-a10d-746e957582a0, returning value {'mode_confirm': 'walk', 'purpose_confirm': 'exercise', 'replaced_mode': 'bus'}
2021-07-20 10:54:54,972:DEBUG:4457897408:In placeholder_predictor_demo: ound 85 for user 1d292b85-c549-409a-a10d-746e957582a0, returning value {'mode_confirm': 'skateboard', 'purpose_confirm': 'exercise', 'replaced_mode': 'same_mode'}
```

Final result was a combination of real and placeholder

```
2021-07-20 10:54:55,182:DEBUG:4457897408:Found real prediction Labelprediction({'trip_id': ObjectId('60f70def49a6d6de8e8ef803'), 'algorithm_id': 4, 'prediction': [{'labels': {'mode_confirm': 'not_a_trip', 'purpose_confirm': 'home', 'replaced_mode': 'not_a_trip'}, 'p': 1.0}], 'start_ts': 1612036947.149, 'end_ts': 1612037455.46}), using that preferentially
2021-07-20 10:54:55,286:DEBUG:4457897408:No real prediction found, using placeholder prediction Labelprediction({'trip_id': ObjectId('60f70def49a6d6de8e8ef807'), 'algorithm_id': 5, 'prediction': [{'labels': {'mode_confirm': 'walk', 'replaced_mode': 'same_mode', 'purpose_confirm': 'meal'}, 'p': 0.9517281073260278}], 'start_ts': 1612132424.3783095, 'end_ts': 1612133524.871})
```

- Reset confirmed trips

```
all inferred trips 215
all confirmed trips 0
confirmed trips with inferred labels 0
confirmed trips without inferred labels 0
confirmed trips with expectation 0
```

- Reran the pipeline

```
all confirmed trips 215
confirmed trips with inferred labels 215
confirmed trips without inferred labels 0
confirmed trips with expectation 215
```

shankari commented:
@GabrielKS, while testing the new placeholder, labeling code (https://github.com/corinne-hcr/e-mission-server/pull/2/commits/f254af9eafe29be59e6ca4ac8e23d77807dccd8d), I found what appears to be an error in the pipeline. The expectation trips are regenerated on every run of the intake pipeline **even when they are not reset in the middle**. I'm pretty sure that this due to some weird constant thing, but I don't have the time to check.

Are you able to debug?

To reproduce, start with a database that has confirmed trips

```
storage not configured, falling back to sample, default configuration
Connecting to database URL localhost
All inferred trips 215
Inferred trips with inferences 215
All expected trips 215
Expected trips with inferences 215
Found confirmed trip with matching inferred trip, without user labels
Found confirmed trip with matching inferred trip, without user labels
Found confirmed trip with matching inferred trip, without user labels
Found confirmed trip with matching inferred trip, without user labels
Found confirmed trip with matching inferred trip, without user labels
Found confirmed trip with matching inferred trip, without user labels
Found confirmed trip with matching inferred trip, without user labels
all inferred trips 215
all confirmed trips 215
confirmed trips with inferred labels 215
confirmed trips without inferred labels 0
confirmed trips with expectation 215
```

Reset only the confirmed trips and rerun the pipeline.

```
$ ./e-mission-py.bash bin/debug/reset_partial_label_testing.py -c
storage not configured, falling back to sample, default configuration
Connecting to database URL localhost

$ ./e-mission-py.bash bin/debug/intake_single_user.py -u 1d292b85-c549-409a-a10d-746e957582a0
...
2021-07-20T12:00:32.769625-07:00**********UUID 1d292b85-c549-409a-a10d-746e957582a0: inferring transportation mode**********
2021-07-20T12:00:32.793987-07:00**********UUID 1d292b85-c549-409a-a10d-746e957582a0: inferring labels**********
2021-07-20T12:00:32.816332-07:00**********UUID 1d292b85-c549-409a-a10d-746e957582a0: populating expectations**********
2021-07-20T12:00:33.419427-07:00**********UUID 1d292b85-c549-409a-a10d-746e957582a0: creating confirmed objects **********
2021-07-20T12:00:51.498790-07:00**********UUID 1d292b85-c549-409a-a10d-746e957582a0: storing views to cache**********
```

the number of expected trips, and thus, the number of confirmed trips, doubles. (I actually ran the pipeline twice so it tripled in my case).

```
 ./e-mission-py.bash bin/debug/label_stats.py -u 1d292b85-c549-409a-a10d-746e957582a0
All inferred trips 215
Inferred trips with inferences 215
All expected trips 645
Expected trips with inferences 645
...
all inferred trips 215
all confirmed trips 645
confirmed trips with inferred labels 645
confirmed trips without inferred labels 0
confirmed trips with expectation 645
```

GabrielKS commented:
I'll look at this now.

shankari commented:
workaround for https://github.com/corinne-hcr/e-mission-server/pull/2#issuecomment-883624475 is to delete all expected trips when we delete confirmed trips. With that workaround, I should be able to push this to staging.

@GabrielKS can you also briefly review https://github.com/corinne-hcr/e-mission-server/commit/f254af9eafe29be59e6ca4ac8e23d77807dccd8d since you wrote the original code?

shankari committed:
Workaround to delete all expected trips when we delete confirmed trips

Before fix:
https://github.com/corinne-hcr/e-mission-server/pull/2#issuecomment-883624475

After fix:

- Before reset
```
Inferred trips with inferences 215
All expected trips 430
Expected trips with inferences 430
all inferred trips 215
all confirmed trips 215
confirmed trips with inferred labels 215
confirmed trips without inferred labels 0
confirmed trips with expectation 215
```

- After reset

```
$ ./e-mission-py.bash bin/debug/reset_partial_label_testing.py -c
storage not configured, falling back to sample, default configuration
Connecting to database URL localhost
{'n': 430, 'ok': 1.0}
{'n': 215, 'ok': 1.0}
{'n': 1, 'ok': 1.0}

storage not configured, falling back to sample, default configuration
Connecting to database URL localhost
All inferred trips 215
Inferred trips with inferences 215
All expected trips 0
Expected trips with inferences 0
...
confirmed trips with inferred labels 0
confirmed trips without inferred labels 0
confirmed trips with expectation 0
```

- After re-run

```
All expected trips 215
Expected trips with inferences 215
all inferred trips 215
all confirmed trips 215
confirmed trips with inferred labels 215
confirmed trips without inferred labels 0
confirmed trips with expectation 215
```

shankari commented:
Workaround committed (https://github.com/corinne-hcr/e-mission-server/pull/2/commits/d301544902d1810a7720b856a8f4993c96dfcc16). @GabrielKS, feel free to push changes in the next 10 mins. After that, I want to push this and actually re-run the pipeline.

GabrielKS commented:
I have no changes to push at the moment.

shankari commented:
About to reset (dun-dun-dun!!)

shankari commented:
First round of reset complete
Ran the pipeline
Second round of reset complete
Pipeline will run at :15

@GabrielKS, can you test in ~5 mins?
Then at least we can do a demo in which you share your screen.

corinne-hcr commented:
Should I pull put the `map labels` to the beginning stage(before the the 1st round of clustering)? Would it be faster?
I think the reason for calculating the distance matrix multiple times is from the 1st round. Some trips need to compare with almost every other trip to see if they can be in the same bin.

shankari commented:
@corinne-hcr can you please focus on getting the unit tests working? I want to have some backup before significant refactoring. I will ask you if I have other questions or suggestions for immediate changes.

corinne-hcr commented on emission/analysis/modelling/tour_model/evaluation_pipeline.py:
```diff
@@ -128,16 +133,20 @@ def test(data,radius,low,dist_pct,kmeans):
+    for a, user in enumerate(all_users):
+        logging.info(f"Starting evaluation for {user}")
         df = pd.DataFrame(columns=['user','user_id','percentage of 1st round','homogeneity socre of 1st round',
                                    'percentage of 2nd round','homogeneity socre of 2nd roun','scores','lower boundary',
```
I have a typo in this line.
Should be `homogeneity socre of 2nd round`

shankari replied:
go ahead and submit a commit to that

corinne-hcr replied:
Should I merge your PR first and then submit a commit? Or just submit the commit?

GabrielKS committed:
Refactor to allow for multiple ensemble options

+ Add the ability to select an ensemble function like primary_algorithms does for normal inference algorithms

GabrielKS committed:
Refactor prediction, ensemble algorithms out of pipeline.py

+ Prediction algorithms now go in inferrers.py
+ Ensemble algorithms now go in ensembles.py

GabrielKS committed:
Edit test to account for refactoring

+ Just needed to update the fully qualified names of the predictors
- Note that the ensemble test is weak enough that it did not need to be
  changed -- this may or may not be something to work on in the future

GabrielKS committed:
Fix accidental ensemble change, fix AlgorithmTypes bug

+ In a previous commit, I accidentally changed the selected ensemble; here, I change it back
+ AlgorithmTypes was misnumbered -- both PLACEHOLDER_2 and PLACEHOLDER_3
  had values of 3 -- here, I fix that

shankari commented:
@GabrielKS @corinne-hcr 

The steps to generate the inferences currently are (all files in `emission/analysis/modelling/tour_model/`)
- run `evaluation_pipeline.py` - this runs cross-validation and generates the results for various turning parameters
- run `build_save_model.py` - this generates a model per user with the tuned parameters and saves it

Once these generate the saved models (e.g. you have files called `models_*` and `locations_*`) for at least some users, you can run the pipeline, which will invoke a method from `load_and_predict.py`

shankari committed:
Add a hack to allow us to reset the URL that the code connects to

The database code has currently been optimized for a 1:1 mapping between the
server and the database, where the server needs to make multiple intensive
queries to the database, for example, to read trajectory information from the
database for multiple users at a time, for the post-processing.

This has led to design decisions in which we cache the database connection and
re-use it, to avoid overloading the server with multiple queries.

We are now getting some requests for federated data.

A concrete example is
https://github.com/e-mission/e-mission-eval-private-data/pull/28/commits/952c476dec745eb66477fed8d699f1bab85d84c0
where we wanted to compare the characteristics of multiple datasets
(https://github.com/e-mission/e-mission-eval-private-data/pull/28)

An upcoming example would be to "roll up" multiple dashboard deployments (e.g.
from individual cities) into a program level dashboard. We anticipate that
these will be low, volume, intermittent accesses to generate analyses and
metrics.

The long-term fix is probably to create a FederatedTimeseries (similar to the
AggregateTimeseries that merges data across users). But for now, we just
implement a hack to reset the connection and reconnect it to a different URL.
This means that we cannot access all databases in parallel, we will need to
access them serially. But for the current use case, that is sufficient since we
can concatenate all the data and work with it later.

shankari committed:
Use the trip start and end location directly

Instead of looking up the place and getting it instead.

This has two advantages over the current implementation:
1. We don't have to make 2 separate database calls for each trip
    Note that we compute an nxn distance matrix, so this is likely to be a
    substantial savings
2. We can pass in a in-memory trip list. That makes it easier to write unit
    tests, and to use alternate load methods (e.g. for working with federated data
    https://github.com/e-mission/e-mission-eval-private-data/pull/28/commits/952c476dec745eb66477fed8d699f1bab85d84c0

@corinne-hcr reported that the place location and the trip start/loc locations
are not identical. We don't have unit tests to verify this (alas!) but the top
level results are not changed significantly.

So the ROI seems high enough; we are going ahead with this change.
https://github.com/e-mission/e-mission-server/pull/826#issuecomment-885714205

shankari committed:
Create a function to expand the userinput into separate dataframe columns

If the labels are simple KV pairs, the result columns are scalar.
Currently duplicated from:
https://github.com/e-mission/em-public-dashboard/blob/main/viz_scripts/scaffolding.py
but without the ipython dependencies and the assumption that there will always
be three entries in the userinput. We removed the assert to handle this assumption.

TODO: Consider replacing with pandas.io.json.json_normalize

shankari committed:
Support multiple usage patterns + return labels

This commit has two main components.

The simpler component is to simply pull out the trip filteration code into a
separate function that we can invoke externally instead of embedding it into
the init. This allows us to experiment with the clusters both with and without
filtering. Note that, if we want to be consistent with sklearn, we should
convert this to a set of parameters instead.

The second component is a function to generate "label" for all the trips that
are passed in.

This is pulled out into a separate function because of the tricksiness of
filtering followed by binning, in which the indices in the bins are from the
filtered trips and not the complete set of trips passed in.

So if we passed in 20 trips (indices = `range(20`)
and the following trips were filtered out:
`[0, 5, 10, 15]`
the bins would be

```
[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]]
```

However, using those indices directly will return labels
`[0,0,0,0,1,1,1,1...]`. So trip 5 would get label of 1 although it was actually
filtered out.

To fix this, we store trip arrays of the various stages of the algorithm, and,
when appropriate, map from bin indices -> all trip indices before setting the
labels.

Testing done:
Added tests for generating labels for various combinations of filtering, binning and cutoffs.
All tests pass.

shankari committed:
Create a fit() method similar to sklearn + fix corner cases

Now that there are potentially multiple steps to run for the similarity code,
we pass in all the parameters while creating the modeler and write a separate
`fit()` method that invokes the steps according to the parameters and saves the
labels.

So the usage pattern is now:

```
curr_sim = similarity(trips,...params)
curr_sim.fit()
... = curr_sim.labels_
```

I briefly considered moving to an even closer match

```
curr_sim = similarity(...params)
curr_sim.fit(trips)
... = curr_sim.labels_
```

But that would involve more restructuring that I am not comfortable doing
without proper unit tests (@corinne-hcr)

This also includes a few corner cases:
- only one bin
- no filtered trips
found during testing.

I've added a unit test for the "no filtered trips" case since we have unit
tests for that, but there are no tests for the filtering, so I've just made the
code change.

With these changes, I can generate results for all mini-pilot users who have
more than 14 trips.
